### DISCUSSION TITLE
How to use ts-node ESM with node modules?

### DISCUSSION THREAD

**Discussion by KaustubhBadrike-GEP** [TypeStrong/ts-node#1321] [Q&A] [answered]
I am trying to make a simple script that reads a files from one location and processes it. Here is what I have so far, having followed #1007:

Node: v12.22.1
```
my-app
├── .eslintrc.cjs
├── .gitignore
├── in
│   └── given.txt
├── node_modules
├── out
├── package-lock.json
├── package.json
├── src
│   ├── helper.ts
│   └── main.ts
└── tsconfig.json
```
package.json
```json
{
  "name": "my-app",
  "version": "1.0.0",
  "type": "module",
  "scripts": {
    "lint": "eslint ./src --ext .js,.jsx,.ts,.tsx --max-warnings=0",
    "start": "node --experimental-specifier-resolution=node --experimental-modules --no-warnings --loader ts-node/esm ./src/main.ts"
  },
  "author": "Kaustubh Badrike",
  "devDependencies": {
    "@types/node": "^15.3.0",
    "@typescript-eslint/eslint-plugin": "^4.24.0",
    "@typescript-eslint/parser": "^4.24.0",
    "eslint": "^7.26.0",
    "ts-node": "^9.1.1",
    "typescript": "^4.2.4"
  }
}
```
tsconfig.json
```jsonc
{
  "compilerOptions": {
    "target": "es5",                                /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
    "module": "ESNext",                           /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
    
    /* Strict Type-Checking Options */
    "strict": true,                                 /* Enable all strict type-checking options. */
    
    /* Module Resolution Options */
    "moduleResolution": "node",                  /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    "esModuleInterop": true,                        /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
    
    /* Advanced Options */
    "skipLibCheck": true,                           /* Skip type checking of declaration files. */
    "forceConsistentCasingInFileNames": true        /* Disallow inconsistently-cased references to the same file. */
  }
}
```
main.ts
```ts
import { readFileSync } from "fs";

const contents = readFileSync('../in/given.txt');
```

However, when I try `npm run start`, I get the following
```
TypeError [ERR_INVALID_RETURN_PROPERTY_VALUE]: Expected string to be returned for the "format" from the "loader getFormat" function but got type 
object.
    at Loader.getFormat (internal/modules/esm/loader.js:125:13)
    at Loader.getModuleJob (internal/modules/esm/loader.js:247:20)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:49:17)
    at async Promise.all (index 0)
    at link (internal/modules/esm/module_job.js:53:9)
```

The error does not occur if I remove the import from fs. Importing items from helper.ts works as well.

Another, possibly shorter way to reproduce:

1. Create a copy of [ts-node/tests/esm-node-resolver at 9d05cb684fc3a6e492832100a125ea07d1cc98c5 · TypeStrong/ts-node](https://github.com/TypeStrong/ts-node/tree/9d05cb684fc3a6e492832100a125ea07d1cc98c5/tests/esm-node-resolver)
2. Remove all `if` blocks from .ts files.
3. Remove biff.jsx and all transitive references.
3. Run `npm i -D ts-node typescript @types/node`.
4. Create a copy of [libfoo](https://github.com/TypeStrong/ts-node/tree/9d05cb684fc3a6e492832100a125ea07d1cc98c5/tests/esm-node-resolver/node_modules/libfoo) in node_modules, as it is overwritten during install.
5. Run `node --loader ts-node/esm --experimental-specifier-resolution=node index.ts` as specified in [ts-node/index.spec.ts at 6084866b8041f838b17bd86158962f2fdf63cc14 · TypeStrong/ts-node](https://github.com/TypeStrong/ts-node/blob/6084866b8041f838b17bd86158962f2fdf63cc14/src/test/index.spec.ts#L1198). Everything should work without errors.
6. Change the content of index.ts to following
```ts
import { foo } from './foo';
import { bar } from './bar';
import { baz } from './baz';
import { libfoo } from 'libfoo';
import { readFileSync } from 'fs';

console.log(`${foo} ${bar} ${baz} ${libfoo}`);
const content = readFileSync('package.json', { encoding: 'utf8' });
console.log(content);
```
7. Run `node --loader ts-node/esm --experimental-specifier-resolution=node index.ts` again. It fails.

**What do I need to do to get it to work?**

[Clone on StackOverflow](https://stackoverflow.com/q/67600063/11928194)

**Comment by felipeplets** [comment]
@KaustubhBadrike-GEP I could reproduce your issue on NodeJS 12.22.1. I can also confirm that it work properly on NodeJS 14.16.

Is updating your NodeJS version a possibility? 

Keep in mind that NodeJS support to ESM is experimental for a long time and changes are usually expected in this phase.

Also, related to that fact you are using a version of NodeJS (12.X) that is approaching the [end of it's maintenance life](https://nodejs.org/en/about/releases/), my recommendation is that you update your NodeJS to at least version 14.x either to solve this issue and for you to benefit on the latest development and guarantee long term support.

**Reply by KaustubhBadrike-GEP** [reply]
Unfortunately, I can't update my NodeJS version. The script I'm working on is to be used in a development team, many of whom have Node 12 installed.

**Reply by cspotcode** [reply]
The error code you are seeing was mentioned here: https://github.com/TypeStrong/ts-node/issues/1007#issuecomment-689860997
The user who reported it explained how they were able to avoid the issue here: https://github.com/TypeStrong/ts-node/issues/1007#issuecomment-690782165

The `--experimental-specifier-resolution` used to have a different name: `--es-module-specifier-resolution`

The reporter also sent us this PR: #1122 which was merged and released in [v9.1.0 / v9.1.1](https://github.com/TypeStrong/ts-node/releases)

**Reply by cspotcode** [reply]
I see that you created a reproduction using our own unit test.  Thank you!  I'll try to reproduce on my machine.

**Reply by cspotcode** [reply]
Found it; should be fixed by #1332.

However, we may choose not to fix ESM breakages in the future.  The experimental nature means that node can choose to change things in a new node release, we'll need to change our loader to match, and it might create an incompatibility with older node versions.  This would create a maintenance headache for us, since node has not committed to a stable API, and we can't necessarily support multiple API varieties.

**Reply by max-hk** [reply]
ESM is now stable on Node.js as of v16.x

https://nodejs.org/docs/latest-v16.x/api/esm.html#esm_modules_ecmascript_modules

**Reply by cspotcode** [reply]
@maxloh not the loaders.
https://nodejs.org/docs/latest-v16.x/api/esm.html#esm_loaders

**Comment by cspotcode** [comment, marked as answer]
Fixed by #1332.
